### ISSUE TITLE
[ENH] Better integration of NeuralFOrecaster

### ISSUE THREAD

**Issue by corradomio** [sktime/sktime#6656]
In sktime there are no models based on NN.
I found 3 libraries:

1) Darts
2) Python Forecasting
3) NeuralForecaster
4) Flow Forcast (2022)
5) tsai (2024)

The library with the highest number of models is NeuralForecast.

However, the library's philosophy is orrible: a FLAT LONG list of parameters for each model.

I think the integration with 'sktime' can be the opportunity to organize the parameters in "logical" groups:

1) TS specific: windows_length, prediction length, etc
2) preprocessing: the scaler
3) model: parameters model's specific (number of layers, activation function, etc)
4) engine: learning rate, loss function, optimizer, ...
5) training: number of epochs, batch size, ...

It is MANDATORY doesn't specify parameters it is possible to infer from the data passed in input: 
because sktime's phylosophy is "fit(y, X)" "predict(fh, X)", it is possible to infer 'stat_exog_list', 'hist_exog_list', 'futr_exog_list, 'freq'.

Note: I started a new integration because I don't like how the library is currently used: the library is used AS IS.

**Comment by yarnabrina** [comment]
@corradomio not sure if you are using an old version of sktime or not, but we do haveadapter for `neuralforecast`, and RNN and LSTM has been interfaced using this.

If you are interested in interfacing more models, contributions towards #6459 are much welcome.

**Comment by corradomio** [comment]
I am using the latest and the dump from github

**Comment by corradomio** [comment]
I am using the latest and the dump from github
rnn and lstm are 2 very simple models that everyone is able to implement and integrate. 

The problem in not on these two simple models, but in the more complex and advanced ones: nhits, beats, based on transformer, etc..

There is a rich collection of models and new ones are proposed every week.
.
This could be an important improvement in the competition with other python libraries:
 to have a rich/well organized/implemented inhouse NN models

(based on pytorch & skorch - instead than pytorch lightning)

**Comment by fkiraly** [comment]
> In sktime there are no models based on NN.

You mean native implementations of forecasters, right?

Could you give your opinion on the following:

* do you agree or disagree with this statement: we should not duplicate if other libraries have reasonable implementations.
* having said that, I agree that not all implementations are production grade, though interfaces are a quicker win (and less resource intensive) than re-implementations.
* @fnhirwa was doing a prioritization on de-novo implementations. I have opened an issue to collect:
    * https://github.com/sktime/sktime/issues/6660
    * can you add anything you would like to see to the wishlist? Since we are well-resourced in terms of contributions at the moment, this gives it a good chance we implement over the next months!

**Comment by fkiraly** [comment]
> However, the library's philosophy is orrible: a FLAT LONG list of parameters for each model.

Hm, why do you consider this to be bad? Can you explain?

This is how parameters are typically managed in `sklearn`.

Can you point to an alternative design - actual or speculative code of how it could look like?

**Comment by fkiraly** [comment]
> Note: I started a new integration because I don't like how the library is currently used: the library is used AS IS.

Integration with the `sktime` API? A pull request would be much appreciated in this case!

**Comment by fkiraly** [comment]
Re the general API points, @fnhirwa is also collating refactor/rework suggestions in this issue:

https://github.com/sktime/sktime/issues/6641

Feel free to comment there, @corradomio!

**Comment by corradomio** [comment]
Currently I am integrating NeuralForecast & Darts in sktime, following the sktime phylosophy
Howver, NeuralForecast is SLOW, VERY, VERY, VERY slow.

The integration is PLAIN, just a SIMPLE wrapper to convert sktime's fit/predict into neuralforecat's fit/predict, 
and some information (stat_exog_list, hist_exog, futr_exog_list) are inferred directly from X/y
I have a MY implementation of several NN for TS, based on pytorch and skorch, and they are NOT so slow.

I am using excellent hardware (a LOT of memory, a LOT of memory in GPU, a LOT of cores in CPU, a LOT of cores in GPU, a LOT of ...) , this behaviour is not reasonable.

I have to check IF it is a problem with pytorch lightning configuration

**Comment by fkiraly** [comment]
do you know which layer the problem is coming from? Is it `sktime` or sth else?

I suppose it's the python equivalent of "catching fire" ...
See https://github.com/sktime/sktime/issues/6660#issuecomment-2187088972

**Comment by fkiraly** [comment]
Re `neuralforecast`, are you using the adapter that @yarnabrina has written? Afaik he is also using it in production and he has not complained so far...

**Comment by yarnabrina** [comment]
> Afaik he is also using it in production and he has not complained so far...

I am not using `neuralforecast` in production currently. It has too many weird specifics issues that blocks me to use reliably. I put this exploration on hold for some other priority works, which unfortunately affects @pranavvp16 as well since I put on hold the Auto-adapter works.

---

> Howver, NeuralForecast is SLOW, VERY, VERY, VERY slow.
> 
> The integration is PLAIN, just a SIMPLE wrapper to convert sktime's fit/predict into neuralforecat's fit/predict, and some information (stat_exog_list, hist_exog, futr_exog_list) are inferred directly from X/y I have a MY implementation of several NN for TS, based on pytorch and skorch, and they are NOT so slow.

The adapter I wrote was intended to make those models available to `sktime` users in the same unified format they can use other models. So, yes that's a simple wrapper and I do not plan to make that one any different, unless you are saying that you found everything else (dataset volume, available CPU, available memory, etc.) unchanged, using `neuralforecast` package is significantly faster than the adapter existing in `sktime`.

@corradomio If you have your own implementation of the same algorithms that are there in `neuralforecast` and those are faster, please contribute, either in `sktime` or in `neuralforecast`, or in your own package. And please do share a benchmark showing the time difference between direct `neuralforecast` use vs use through `sktime` adapter, contrasting both with your own implementation.

**Comment by corradomio** [comment]
A sktime adapter to use NF is SIMPLE/stupid: it reorganizes y, X in the following way:

1. it converts sktime 'y' data into a dataframe with columns "unique_id" (an integer, used to identify different TS in a multi TS's dataframe), "y" (the target), "ds" (the timestamp). Ok, this requires time, BUT the process is one-time
2. from X, if available it retrieved the values to populate 'stat_exog_list', 'hist_exog_list', 'futr_exog_list' (this is the INFERRED part)
3. it concatenates [y, X]
4. THE END
.
Ok, it is necessary to create the instance of the model, to call "NF.fit(df)", then to call "NF.predict() -> df" to obtain the prediction in the "prediction windows" specified in configuration.
There is another little trick to apply, when "fh" is LONGER than the "prediction windows", it is necessary to iterate multiple time on "NF.predict()".
.
This to say, IF the model is SLOW, IT IS NOT RESPONSIBILITY of the wrapper BUT it is responsibility of model ITSELF.
I suppose:

1. it is a problem with Lighting configuration (simple to resolve, it is enough to use better configuration parameters)
2. it is a problem in the NN model: this is more difficult to resolve because a NN model for TS is enough "intricated"
3. it is a problem on HOW the TS data is converted in a format specific for the model
.
You have to remember that a NN model for TS is HOWEVER a TABULAR model, than the TS must be converted in tabular format
.
Analyzing in progress ...
